merlin/util/decoding.py

```diff
@@ -167,85 +165,86 @@ def extract_barcodes_with_index(
                 specified barcode index
         """
         properties = measure.regionprops(
-                measure.label(decodedImage == barcodeIndex),
-                intensity_image=pixelMagnitudes,
-                cache=False)
-        dList = [self._bc_properties_to_dict(
-            p, barcodeIndex, fov, distances, pixelTraces, zIndex,
-            globalAligner, segmenter
-        ) for p in properties
-            if self._position_within_crop(
-                p.centroid, cropWidth, decodedImage.shape)
-            and p.area >= minimumArea]
-        barcodeInformation = pandas.DataFrame(dList)
-
-        return barcodeInformation
-
-    @staticmethod
-    def _position_within_crop(position: np.ndarray, cropWidth: float,
-                              imageSize: Tuple[int]) -> bool:
-        if len(position) == 2:
-            return cropWidth < position[0] < imageSize[0] - cropWidth \
-                    and cropWidth < position[1] < imageSize[1] - cropWidth
+            measure.label(decodedImage == barcodeIndex),
+            intensity_image=pixelMagnitudes,
+            cache=False)
+        columnNames = ['barcode_id', 'fov', 'mean_intensity', 'max_intensity',
+         'area', 'mean_distance', 'min_distance',
+         'x', 'y', 'z', 'global_x', 'global_y', 'global_z',
+         'cell_index']
+        if len(pixelTraces.shape) == 3:
```

Comment:
Since we have so many if statements that all check if the data is 3D or 2D, we could make it clearer by defining a variable is3D = len(pixelTraces.shape) == 3 and then using this in the if statements.

Reply:
For the 3D Decoding, is it primarily just allowing pixels to link across Z in the measure.label call? I'm wondering if you want to maintain support for this, as opposed to just using 2D decoding and then cleaning up using the z plane duplicate removal I pushed recently.

Reply:
We should keep support for this since when the z step size is much smaller than 1.5 microns it is useful to connect the voxels in three dimensions to increase the voxel count of correct barcodes for filtering.

Reply:
I agree that could be true, but it's not clear to me it would actually increase separation between foreground and background spots in terms of area, have you found this to be true experimentally?

Reply:
I don't have evidence that the increased voxel count results in improved filtering but it is different enough that I expect the two strategies give different results. Since some users are currently using the 3d decoding we should keep it in place, but I expect that removing the duplicates afterward is much more memory efficient than the current 3d decoding, so if the performance isn't comprised your new strategy is favorable.

Reply:
That sounds good, I agree